src/couch/src/couch_ejson_size.erl

```diff
@@ -53,3 +62,130 @@ ejson_size(false) ->
+        ?horizontal_tab -> utf8_string_size(Rest, Acc + 2);
+        ?newline -> utf8_string_size(Rest, Acc + 2);
+        ?formfeed -> utf8_string_size(Rest, Acc + 2);
+        ?carriage_return -> utf8_string_size(Rest, Acc + 2);
+        NeedEscape when NeedEscape =< 16#001F -> utf8_string_size(Rest, Acc + 6);
+        _Else -> utf8_string_size(Rest, Acc + 1)
+    end;
+%% 10000 .. 10000007F
+utf8_string_size(<<2#1111:4, _:28/bitstring, Rest/binary>>, Acc) ->
+    utf8_string_size(Rest, Acc + 4);
+%% 0800 .. FFFF
+utf8_string_size(<<2#111:3, _:21/bitstring, Rest/binary>>, Acc) ->
+    utf8_string_size(Rest, Acc + 3);
+%% 0080 .. 07ff
+utf8_string_size(<<2#11:2, _:14/bitstring, Rest/binary>>, Acc) ->
+    utf8_string_size(Rest, Acc + 2);
+utf8_string_size(<<>>, Acc) ->
+    Acc.
+
+
+-ifdef(TEST).
+
+-include_lib("eunit/include/eunit.hrl").
+
+proper_test_() ->
+    PropErOpts = [
+        {to_file, user},
+        {max_size, 5},
+        {numtests, 1000}
+    ],
+    properties_tests(properties(), 30, PropErOpts).
+
+properties() ->
+    Exported = ?MODULE:module_info(exports),
+    PropertiesFilter = fun({FunName, Arrity}) ->
+        Arrity =:= 0 andalso lists:prefix("prop_", atom_to_list(FunName))
+    end,
+    [F || {F, _} = Export <- Exported, PropertiesFilter(Export)].
+
+properties_tests(Props, Timeout, Opts) ->
+    [property_test(Name, Timeout, Opts) || Name <- Props].
+
+property_test(Name, Timeout, Opts) ->
+    {atom_to_list(Name),
+        {timeout, Timeout,
+            ?_assert(proper:quickcheck(?MODULE:Name(), Opts))}}.
+
+%% Generators
+
+json_object() -> ?LAZY({json_dict()}).
+
+json_dict() ->
+    ?LET(L, list({json_string(), json_value()}), L).
+
+json_value() ->
+    ?LAZY(
+    frequency([
+        {2, json_array()},
+        {3, null()},
+        {4, boolean()},
+        {5, number()},
+        {6, json_object()},
+        {6, json_string()}
+    ])).
+
+json_array() ->
+    ?LAZY(list(json_value())).
+
+json_string() ->
+    json_string(<<>>).
+
+json_string(Fragments) ->
+    frequency([
+        {1, ?LET(Generated, Fragments, iolist_to_binary(Generated))},
+        {50, ?LAZY(json_string([string_fragement() | Fragments]))}
+    ]).
+
+string_fragement() ->
+    frequency([
+        {1, integer(16#0000, 16#001F)}, %% \u+
+        {10, list(oneof(
+            [
+                ?quotation_mark,
+                ?reverse_solidus,
+                ?solidus,
+                ?backspace,
+                ?horizontal_tab,
+                ?newline,
+                ?formfeed,
+                ?carriage_return
+        ]))},
+        {50, utf8()}
+    ]).
+
+null() ->
+    null.
+
+
+%% Properties
+
+prop_string_external_size() ->
+    ?FORALL(String, json_string(), begin
+        byte_size(jiffy:encode(String)) =:= utf8_string_size(String)
+    end).
+
+prop_number_external_size() ->
+    ?FORALL(EJson, number(), begin
+        byte_size(jiffy:encode(EJson)) =:= ejson_size(EJson)
```

Comment:
The goal is not to be equal to jiffy's encoded size in general but to provide a reasonable minimal encoding for all the possible encoder customer's might use. Think Python, Java, Javascript (node), C++, PHP etc.

We can't ask a customer to install jiffy to test their document sizes. But if we can tell them "whatever language's json encoder you use, if you measure its output size it should be accepted by our system".